### PR TITLE
Refactor TraceQL Fetch Layer

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Refactors the TraceQL fetch layer in preparation for adding the select and by queries. Primarily rewrites the concept of "filter" to be a second pass on the same set of iterators. This required adding "fake" intrinsics for the required metadata.

Additionally this rewrite exposed an issue we'll need to deal with in the fetch layer. Currently the `requireAtLeastOneMatch` optimizations count the number of conditions vs the number of attributes on the span. However:

- span id/start/end are not added to attributes
- kind is added to "count" spans if there are no other required iterators at the span level. kind is then added to the attributes even if there is no condition requesting it
- in vParquet requesting duration translates to retrieving start/end metadata

This PR attempts to keep all tests passing with minimal changes here, but it really should be addressed longer term.
 
~TODO: This change has only been applied to vParquet. Working on vParquet2 now.~

Benchmarks on vParquet vs. main:

```
name                                               old time/op    new time/op    delta
BackendBlockTraceQL/spanAttNameNoMatch-8             30.8ms ±11%    26.6ms ± 1%  -13.58%  (p=0.016 n=5+4)
BackendBlockTraceQL/spanAttValNoMatch-8              34.7ms ± 6%    33.8ms ± 5%     ~     (p=0.310 n=5+5)
BackendBlockTraceQL/spanAttValMatch-8                 424ms ± 6%     404ms ± 3%     ~     (p=0.095 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-8        22.6ms ± 2%    22.3ms ± 1%     ~     (p=0.111 n=5+4)
BackendBlockTraceQL/spanAttIntrinsicMatch-8           391ms ± 7%     422ms ±10%     ~     (p=0.151 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch-8         12.3ms ±10%    11.7ms ± 9%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch-8          48.2ms ± 1%    41.1ms ± 1%  -14.66%  (p=0.029 n=4+4)
BackendBlockTraceQL/resourceAttValMatch-8             554ms ±26%     353ms ± 8%  -36.24%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-8    12.1ms ±16%    10.8ms ±10%     ~     (p=0.056 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch-8       343ms ±11%     319ms ±14%     ~     (p=0.310 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch-8                6.85s ± 9%     6.57s ± 1%     ~     (p=0.413 n=5+4)
BackendBlockTraceQL/mixedValNoMatch-8                 4.29s ±15%     4.14s ± 2%     ~     (p=0.690 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd-8          13.2ms ±18%    12.7ms ± 1%     ~     (p=0.730 n=5+4)
BackendBlockTraceQL/mixedValMixedMatchOr-8            4.38s ± 4%     4.77s ±23%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/mixedValBothMatch-8              12.0ms ±11%    12.2ms ±19%     ~     (p=0.841 n=5+5)

name                                               old speed      new speed      delta
BackendBlockTraceQL/spanAttNameNoMatch-8            433MB/s ±11%   499MB/s ± 1%  +15.26%  (p=0.016 n=5+4)
BackendBlockTraceQL/spanAttValNoMatch-8             577MB/s ± 6%   592MB/s ± 5%     ~     (p=0.310 n=5+5)
BackendBlockTraceQL/spanAttValMatch-8              42.5MB/s ± 6%  44.5MB/s ± 3%     ~     (p=0.095 n=5+5)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-8       593MB/s ± 2%   602MB/s ± 1%     ~     (p=0.111 n=5+4)
BackendBlockTraceQL/spanAttIntrinsicMatch-8         461MB/s ± 7%   428MB/s ± 9%     ~     (p=0.151 n=5+5)
BackendBlockTraceQL/resourceAttNameNoMatch-8        422MB/s ±10%   441MB/s ± 9%     ~     (p=0.421 n=5+5)
BackendBlockTraceQL/resourceAttValNoMatch-8         385MB/s ±15%   468MB/s ± 1%  +21.51%  (p=0.016 n=5+4)
BackendBlockTraceQL/resourceAttValMatch-8           397MB/s ±27%   598MB/s ± 7%  +50.51%  (p=0.008 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-8  71.0MB/s ±15%  79.5MB/s ± 9%     ~     (p=0.056 n=5+5)
BackendBlockTraceQL/resourceAttIntrinsicMatch-8     585MB/s ±11%   632MB/s ±13%     ~     (p=0.310 n=5+5)
BackendBlockTraceQL/mixedNameNoMatch-8             9.02MB/s ± 9%  9.37MB/s ± 1%     ~     (p=0.413 n=5+4)
BackendBlockTraceQL/mixedValNoMatch-8              14.4MB/s ±13%  14.9MB/s ± 2%     ~     (p=0.730 n=5+5)
BackendBlockTraceQL/mixedValMixedMatchAnd-8         391MB/s ±16%   403MB/s ± 1%     ~     (p=0.730 n=5+4)
BackendBlockTraceQL/mixedValMixedMatchOr-8         47.9MB/s ± 4%  44.6MB/s ±20%     ~     (p=0.548 n=5+5)
BackendBlockTraceQL/mixedValBothMatch-8            71.7MB/s ±11%  70.7MB/s ±16%     ~     (p=0.841 n=5+5)

name                                               old MB_io/op   new MB_io/op   delta
BackendBlockTraceQL/spanAttNameNoMatch-8               13.3 ± 0%      13.3 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValNoMatch-8                20.0 ± 0%      20.0 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttValMatch-8                  18.0 ± 0%      18.0 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-8          13.4 ± 0%      13.4 ± 0%     ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicMatch-8             180 ± 0%       180 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttNameNoMatch-8           5.14 ± 0%      5.14 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch-8            19.3 ± 0%      19.3 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttValMatch-8               211 ± 0%       211 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-8      0.85 ± 0%      0.85 ± 0%     ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicMatch-8         200 ± 0%       200 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedNameNoMatch-8                 61.6 ± 0%      61.6 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValNoMatch-8                  61.6 ± 0%      61.6 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchAnd-8            5.14 ± 0%      5.14 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchOr-8              210 ± 0%       210 ± 0%     ~     (all equal)
BackendBlockTraceQL/mixedValBothMatch-8                0.85 ± 0%      0.85 ± 0%     ~     (all equal)
```

### CODE REVIEW

File: tempodb/encoding/vparquet/block_traceql_test.go [hunk 1]
```diff
@@ -201,6 +201,11 @@ func TestBackendBlockSearchTraceQL(t *testing.T) {
 	}
 
 	for _, req := range searchesThatMatch {
+		if req.SecondPass == nil {
+			req.SecondPass = func(s *traceql.Spanset) ([]*traceql.Spanset, error) { return []*traceql.Spanset{s}, nil }
```
Comment (mdisibio) [inline tempodb/encoding/vparquet/block_traceql_test.go:205]: This looks like the default behavior of `MustExtractFetchSpansRequestWithMetadata` already. Is this still needed?

File: pkg/traceql/storage.go [hunk 2]
```diff
@@ -12,12 +12,37 @@ type Condition struct {
...
+		{NewIntrinsic(IntrinsicTraceRootSpan), OpNone, nil},
+		{NewIntrinsic(IntrinsicTraceDuration), OpNone, nil},
+		{NewIntrinsic(IntrinsicTraceID), OpNone, nil},
+		{NewIntrinsic(IntrinsicTraceStartTime), OpNone, nil},
+		{NewIntrinsic(IntrinsicSpanID), OpNone, nil},
+		{NewIntrinsic(IntrinsicSpanStartTime), OpNone, nil},
+	}
+}
+
 // FilterSpans is a hint that allows the calling code to filter down spans to only
```
Comment (mdisibio) [inline pkg/traceql/storage.go:40]: nit: Update comment

**Review by mdisibio — COMMENTED** [review]
I'm really liking this direction and the flexibility of the SecondPassConditions vs the previous filter.   We reuse Fetch for things like tag lookup and metrics, which has been working out well, and this changes makes it even better.   I need some time to dig deeper but I left some questions.  (We already talked about this a bunch offline so great overall, but I want to give it the level of review that it deserves).

What do you think is driving the benchmark improvements?  Since the input/output should basically be identical?  The only thing I can think of is this fixed some subtle "allconditions-y" column join behavior that wasn't optimized before.

File: pkg/traceqlmetrics/metrics.go [hunk 3]
```diff
@@ -159,13 +159,28 @@ func GetMetrics(ctx context.Context, query string, groupBy string, spanLimit int
...
 	addConditionIfNotPresent(groupByAttr)
 
-	// This filter callback processes the matching spans into the
-	// bucketed metrics.  It returns nil because we don't need any
-	// results after this.
-	req.Filter = func(in *traceql.Spanset) ([]*traceql.Spanset, error) {
+	// Perform the fetch and process the results inside the Filter
+	// callback.  No actual results will be returned from this fetch call,
+	// But we still need to call Next() at least once.
+	res, err := fetcher.Fetch(ctx, *req)
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply pkg/traceqlmetrics/metrics.go:165]: Correct, b/c there is no second pass specified this does not pull any metadata. I think it's one of the bigger wins in this PR. Metrics and tag values don't need to do this weird filter dance for no reason. They now specify the columns they want and only get those back.

[the remark this answers is not in the document]
Comment (joe-elliott) [reply pkg/traceqlmetrics/metrics.go:165]: Correct, b/c there is no second pass specified this does not pull any metadata. I think it's one of the bigger wins in this PR. Metrics and tag values don't need to do this weird filter dance for no reason. They now specify the columns they want and only get those back.

**Comment by joe-elliott** [comment]
> What do you think is driving the benchmark improvements? Since the input/output should basically be identical?

I don't know and haven't dug into. There's a lot of variance in these benchmarks due to their size and running them on my wimpy laptop. I mainly looking for obvious performance degradation. I'd take the improvements with a grain of salt.

File: tempodb/encoding/vparquet/block_traceql_test.go (hunk 1 above)
Comment (joe-elliott) [reply tempodb/encoding/vparquet/block_traceql_test.go:205]: Some of the test cases construct conditions directly and don't use `MustExtractFetchSpansRequestWithMetadata`. Honestly I think i'm going to revert the changes to `MustExtractFetchSpansRequestWithMetadata` and just leave this.

File: pkg/traceqlmetrics/metrics.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdisibio) [reply pkg/traceqlmetrics/metrics.go:165]: 👍 Metrics could actually take advantage of it and put duration/status/groupby in the second pass.  Not in this PR, but sounds great.

File: tempodb/encoding/vparquet/block_traceql_test.go (hunk 1 above)
Comment (mdisibio) [reply tempodb/encoding/vparquet/block_traceql_test.go:205]: Ah that's right.  For OpNone which is outside of the language spec for now, but will be covered by `select()` and `by()` very soon.

File: tempodb/wal/wal_test.go [hunk 4]
```diff
@@ -271,7 +272,7 @@ func testSearch(t *testing.T, e encoding.VersionedEncoding) {
 func TestFetch(t *testing.T) {
 	for _, e := range encoding.AllEncodings() {
 		t.Run(e.Version(), func(t *testing.T) {
-			testFetch(t, e)
+			testFetch(t, vparquet.Encoding{})
```
Comment (mdisibio) [inline tempodb/wal/wal_test.go:275]: typo?

File: tempodb/tempodb_search_test.go [hunk 5]
```diff
@@ -66,7 +66,7 @@ func TestTraceQLCompleteBlock(t *testing.T) {
 	for _, v := range encoding.AllEncodings() {
 		vers := v.Version()
 		t.Run(vers, func(t *testing.T) {
-			testTraceQLCompleteBlock(t, vers)
+			testTraceQLCompleteBlock(t, "vParquet")
```
Comment (mdisibio) [inline tempodb/tempodb_search_test.go:69]: typo?

**Comment by joe-elliott** [comment]
Good catch on the local testing changes. Would have been crappy to merge.

**Pull request merged**